### DISCUSSION TITLE
Osu!Lazer random massive keypress delay during gameplay on macOS

### DISCUSSION THREAD

**Discussion by Whatnoww** [ppy/osu#25518] [Q&A]
### Type

Game behaviour

### Bug description

Osu Lazer on my M1 macbook pro clearly eats my inputs or delays them seemingly at random moments.

Details:
M1 Macbook Pro 13 inch connected to an LG 165hz freesync monitor with freesync enabled. Using tablet + an external keyboard. I have never experienced this issue before. I believe about two stable updates ago this was not an issue? But now it's consistently a problem and makes Lazer unplayable. My keybindings are A and S.

I've gone ahead and recorded my Osu Lazer gameplay with OBS and a keyboard overlay from the OS in the corner.

As you can see in the following image, the cursor is clearly over hitcircle number 4, and in the overlay you can clearly see that A has been pressed (A is displayed in red in the keyboard overlay).

![image](https://github.com/ppy/osu/assets/59415794/68f318da-1583-4408-9ebf-3c28891a0d5b)

Yet a few frames later, it is counted as a full miss, as it seems that the game has detected my keypress once my cursor has left the hitcircle.

![image](https://github.com/ppy/osu/assets/59415794/b7ca9e51-e9a2-4354-9449-134b0de01322)

If I check the replay, this does seem to be in fact what happens. The game seems to be registering my input way later (perhaps when I'm letting go of the key instead of pressing down on it on ocasions?).

If we watch an excerpt of the replay, you can see that the keypress is registered way later, so much later that there's no confusion on what keypress it's registered. It's registered A way later when I'm almost halfway to the next hitcircle. You can see this in the following video at 00:57 and again slower at 01:24 where no input appears close to the first screenshots:

https://youtu.be/a_ro_SLhRXA

I am uploading the replay file here:
[Whatnoww playing Agnete Kjolsrud - Get Jinxed (Tarrasky) [Master] (2023-11-20_14-17).osr.zip](https://github.com/ppy/osu/files/13415454/Whatnoww.playing.Agnete.Kjolsrud.-.Get.Jinxed.Tarrasky.Master.2023-11-20_14-17.osr.zip)

Did anther replay, this time with both the keyboard overlay and osu keyboard overlay enabled:

At 0:31, you can see that the 5th hitcircle is missed entirely, even though going frame by frame, macOS displays it being pressed. Worse thing is that I believe this happens completely at random, which makes it feel pretty bad.

https://youtu.be/DeggmDrZDr0

### Version

2023.1114.1-lazer

### Logs

[database.log](https://github.com/ppy/osu/files/13415618/database.log)
[input.log](https://github.com/ppy/osu/files/13415619/input.log)
[network.log](https://github.com/ppy/osu/files/13415620/network.log)
[performance.log](https://github.com/ppy/osu/files/13415621/performance.log)
[runtime.log](https://github.com/ppy/osu/files/13415622/runtime.log)

**Comment by peppy** [comment]
On macOS, this happens when another app is blocking the OS thread. Let me know when you figure out which app it is (make sure to check things which are running in the menu bar as well).

If it only started recently, you should check apps which you only installed recently, or had large updates recently.

**Reply by Whatnoww** [reply]
Thank you. I will take a look at this Peppy. I'm new to macOS so I am unsure of how to look this up. Looking online doesn't seem to return many results. Should I be looking at activity monitor on another display?

For now I've closed everything I think I possibly can, yet I'm still getting this. Should recording with OBS be safe? That'll help me find the issue if it's fine.

**Comment by Whatnoww** [comment]
I don't think I'm actually able to check for this issue without also making it worse by screen recording. I've done a few test runs including testing with the macOS internal screen recording and seem to run into the same issue regardless, even when I'm not recording. Not sure what further steps to take here, since I think I closed everything I'm allowed to close on macOS and I'm still seeing this issue. There's a chance I could try external capture, but I don't have any hardware capable of capturing 120hz output, so it seems kind of impossible to check for.

I've tried analyzing footage with no screen recording and I still get these phantom delayed clicks with nothing running in the background as far as I am concerned.

I've gone ahead and disabled all background items as well, and I still seem to be getting this issue with just Osu! running on it's own :(

**Reply by peppy** [reply]
Did you miss my message? Simply close one app at a time until you figure which app on your system is causing the issue. Trial and error.

If it happens with "no apps" there's likely something still running in the background or menu bar causing this.

**Reply by Whatnoww** [reply]
Sorry, I had tried this, but had kept on screen recording with the native screen recorder, hence the issue never "disapeared". The culprit seems to be any screen capture at all.

Having any number of apps open doesn't seem to cause any issues, but the second you start screen capturing, these issues pop up, whether the screen capture is done via OBS, Discord, or native screen recording.

Thank you, I will avoid any further screen capture while Osu is open.

**Reply by Whatnoww** [reply]
Regardless I will continue monitoring the situation. I'm unsure why opening all apps then closing them again seems to result in different outcomes from a clean boot. I'll try monitoring the process list to see if any apps persist after closing, or if any new services are running.

**Comment by Whatnoww** [comment]
After having disabled these options I have not had a re-incident of the issue as of since I last updated this thread. Will update if it has solved the issue entirely for me.

![image](https://github.com/ppy/osu/assets/59415794/5ac813e4-dc5e-4c9f-83d6-015a64e9e5b2)

**Reply by Aerodite** [reply]
Tried this and it fixed my keyboard input delay issue as well, thanks!

**Comment by martinledl** [comment]
I tried your solution and it somehow makes it a bit better, not sure if it is a coincidence or not. Still has not solved 100%. When I try to play League of Legends - Flash Funk (Marshmello Remix) - Hard by Pachiru, I consistently have this issue with key delay. No other song caused so much trouble, still sometimes happens even elsewhere. 

Maybe try this song as well, see what it does in your case. Let me know if you find a solution to this!
